src/opower/utilities/smud.py

```diff
@@ -82,99 +82,174 @@ def timezone() -> str:
         username: str,
         password: str,
-        optional_mfa_secret: Optional[str],
-    ) -> str:
+        optional_mfa_secret: Optional[str],  # Not used by SMUD.
+    ) -> None:
         """Login to the utility website and authorize opower."""
-        login_parser = SMUDLoginParser()
+        if (
+            session.cookie_jar.filter_cookies("https://smud.opower.com/ei").__len__()
+            > 0
+        ):
+            return
+
+        smud_login_page_url = "https://myaccount.smud.org/"
+
+        _LOGGER.debug("Fetching SMUD login page: %s", smud_login_page_url)
 
-        async with session.get(
-            "https://myaccount.smud.org",
+        myaccount_response = await session.get(
+            smud_login_page_url,
             headers={"User-Agent": USER_AGENT},
             raise_for_status=True,
-        ) as resp:
-            login_parser.feed(await resp.text())
+        )
+
+        SMUD.log_response(myaccount_response, session)
+
+        # Parse the verification token which will be used during login.
+        # NB: Although the response cookies contain a `__RequestVerificationToken`, it does not match
+        # the one contained in the HTML and results in a 500 server error from SMUD during login.
+        login_parser = SMUDLoginParser()
+        login_parser.feed(await myaccount_response.text())
+        requestVerificationToken = login_parser.verification_token
 
-            assert (
-                login_parser.verification_token
-            ), "Failed to parse __RequestVerificationToken"
+        _LOGGER.debug("Performing SMUD login to: %s", smud_login_page_url)
 
-        await session.post(
-            "https://myaccount.smud.org",
+        # Do the SMUD login which will set cookies in the session.
+        login_response = await session.post(
+            smud_login_page_url,
             data={
-                "__RequestVerificationToken": login_parser.verification_token,
-                #"GenericMessage": "Incorrect username or password",
-                "GenericMessage": "Sorry, you could not be authenticated with the information provided. Please try again.",
-                "LockedMessage": "Account is locked",
-                "ReturnUrl": "",
-                #"OtherPageUrl": "False",
+                "Lang": "en",
                 "UserID": username,
                 "Password": password,
-                #"RememberMe": "true",
+                "__RequestVerificationToken": requestVerificationToken,
             },
             headers={"User-Agent": USER_AGENT},
             raise_for_status=True,
         )
 
-        usage_parser = SMUDUsageParser()
-        #session: aiohttp.ClientSession,
-        async with session.get(
-            #"https://www.pse.com/api/AccountSelector/GetContractAccountJson",
-            "https://myaccount.smud.org/manage/opowerresidential/energyusage",
-            allow_redirects=True,
+        SMUD.log_response(login_response, session)
```

Comment:
For the case when the provided credentials are invalid, it might be a good idea to add something like this:

```suggestion
        SMUD.log_response(login_response, session)
        
        login_response_body = await login_response.text()
        if "could not be authenticated" in login_response_body:
            raise opower.InvalidAuth
```

Otherwise it'll fail with a very opaque message:
```
   @classmethod
    def get_okta_url_from_response_redirect(
        cls, energyusage_response: ClientResponse
    ) -> str:
        """Get the OKTA URL to open next from the last redirect of the previous response."""
        # https://smud.okta.com/login/sessionCookieRedirect
        #   ?token=20111...6QJMn
        #   &redirectUrl=https://smud.okta.com/app/sacramentomunicipalutilitydistrict_opower_1/exk2i...cF0x7/sso/saml
        #       ?RelayState=https://smud.opower.com/ei/app/myEnergyUse
        energyUsageResponseRedirectedFinalUrl = energyusage_response.history[-1].url

        query_parts = parse_qs(energyUsageResponseRedirectedFinalUrl.query_string)

>       return query_parts["redirectUrl"][0]
E       KeyError: 'redirectUrl'

src/opower/utilities/smud.py:226: KeyError
```